Sprint 3/step 1

progpspb opened the pull request.

progpspb committed:
routes and pages

progpspb committed:
protected-route

progpspb committed:
test auth

progpspb committed:
auth

login, register, logout, forgot-password

progpspb committed:
ready for review

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -0,0 +1,63 @@
+
+    const { isError, errMessage } = useSelector(state => state.auth);
+
+    const [ values, setValue ] = useState({email: '', password: ''});
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -0,0 +1,63 @@
+    return (
+        <div className={styles.form_wrapper}>
+            <h2>Вход</h2>
+            <form className={styles.form} onSubmit={handleOnSubmit}>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/components/protected-route/protected-route.jsx:
```diff
@@ -0,0 +1,42 @@
+    }
+    
+    // для неавторизованного пользователя И пользователь авторизован
+    if (!needAuth && user) {
+      const { from } = location.state || { from: { pathname: '/' } };
+      return <Navigate to={from} />;
+    }
+  
+    // для авторизованного пользователя И пользователь не авторизован
+    if (needAuth && !user) {
+      return <Navigate to='/login' state={{ from: location }} />;
+    }
```
Хорошо сделан `ProtectedRoute`

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -4,48 +4,60 @@ import AppHeader from '../app-header/app-header';
 
     const location = useLocation();
     const navigate = useNavigate();
+    const background = location.state && location.state.background;
```
Отлично сделана логика проверки, кликал ли пользователь сам на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку.
* Хорошо сделан `ProtectedRoute`
* Отлично сделана логика проверки, кликал ли пользователь сам на сайте, и показывать ли страницу  с деталями, а не модально окно

но есть некоторые недочеты:



* Ссылки в хэдере теперь не должны перезагружать страницу, так как знаете роутинг
* В консоли ошибки при входе в личный кабинет https://disk.yandex.ru/i/tL8MbqdJYaYt0Q 


**Можно лучше**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше отображать **лоадер** (загрузку) при ожидании ответа от сервера с номером заказа

progpspb committed:
after review 1

checked:
+ hook useForm
+ header links
+ login page console errors
+ clear constructor after order

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
